lightning/mydump/router.go

```diff
@@ -262,115 +278,48 @@ func (p *regexRouterParser) parseFieldExtractor(
-	captures, err := p.extractSubPatterns(t)
-	if err != nil {
-		return errors.Trace(err)
-	}
-
-outer:
-	for _, c := range captures {
-		if isNumber(c) {
-			idx, _ := strconv.Atoi(c)
-			if idx > p.rule.pattern.NumSubexp() {
-				return errors.Errorf("sub pattern capture index '%d' out of range", idx)
-			}
-		} else {
-			for _, n := range p.rule.pattern.SubexpNames() {
-				if c == n {
-					continue outer
-				}
-			}
-			return errors.Errorf("invalid named capture '%s', available names are: %s", c, p.rule.pattern.SubexpNames())
+func (p regexRouterParser) checkSubPatterns(pat *regexp.Regexp, t string, validateFn func(string) error) error {
+	subPats := expandVariablePattern.FindAllString(t, -1)
+	if len(subPats) == 0 {
+		// validate const patterns
+		if err := validateFn(t); err != nil {
+			return err
 		}
 	}
-
-	return nil
-}
-
-func (p *regexRouterParser) extractSubPatterns(t string) ([]string, error) {
-	var res []string
-	remain := t
-	for {
-		idx := strings.IndexByte(remain, '$')
-		if idx < 0 {
-			return res, nil
-		}
-		// $$ is expanding to '$', so just continue
-		if idx < len(remain)-1 && remain[idx+1] == '$' {
-			remain = remain[idx+2:]
+	for _, subVar := range subPats {
+		var tmplName string
+		switch {
+		case subVar == "$$":
 			continue
+		case strings.HasPrefix(subVar, "${"):
+			tmplName = subVar[2 : len(subVar)-1]
+		default:
+			tmplName = subVar[1:]
 		}
-		remain = remain[idx+1:]
-		if len(remain) == 0 {
-			return nil, errors.Errorf("invalid template '%s', cannot end with '$'", t)
-		}
-
-		// parse sub-pattern such as $(pattern) or ${pattern}
-		if remain[0] == '{' {
-			// parenthesis pattern
-			end := strings.IndexByte(remain, '}')
-			if end < 0 {
-				return nil, errors.Errorf("invalid template '%s', named capture starts with '${' should end with '}'", t)
-			}
-			if end == 1 {
-				return nil, errors.Errorf("invalid template '%s', named capture must not be empty", t)
-			}
-			if !isAlphaNumOrUnderline(remain[1:end]) {
-				return nil, errors.Errorf("invalid template '%s', invalid named capture '%s'", t, remain[1:end-1])
+		if number, err := strconv.Atoi(tmplName); err == nil {
+			if number > pat.NumSubexp() {
+				return errors.Errorf("sub pattern capture '%s' out of range", subVar)
 			}
-
-			res = append(res, remain[1:end])
-			remain = remain[end+1:]
-			continue
-		}
-
-		// find the end Index of
-		endIdx := strings.IndexFunc(remain, func(r rune) bool {
-			return r < 48 || (r > 57 && r < 65) || (r > 90 && r != 95 && r < 97) || r > 122
-		})
-		switch {
-		case endIdx == 0:
-			return nil, errors.Errorf("invalid template '%s', named capture can only contain alphabets or numbers", t)
-		case endIdx < 0:
-			res = append(res, remain)
-			remain = remain[:0]
-		default:
-			res = append(res, remain[:endIdx])
-			remain = remain[endIdx:]
+		} else if !slice.AnyOf(pat.SubexpNames(), func(i int) bool {
+			// FIXME: we should use re.SubexpIndex here, but not supported in go1.13 yet
```

Comment:
🤔 we are still using Go 1.13?!

Reply:
yes, in our CI, the go version is still 1.13? The QA team recently started testing updating go version to 1.14/1.15 but  not finished yet.

Reply:
okay 😓